src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,257 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class Http2Connection
+    {
+        private struct Http2StreamWindowManager
+        {
+            public const int StreamWindowUpdateRatio = 8;
+            private static readonly double StopWatchToTimesSpan = TimeSpan.TicksPerSecond / (double)Stopwatch.Frequency;
+
+            private readonly Http2Connection _connection;
+            private readonly Http2Stream _stream;
+
+            private readonly double _windowScaleThresholdMultiplier;
+            private readonly int _maxStreamWindowSize;
+            private bool _windowScalingEnabled;
+
+            private int _deliveredBytes;
+            private int _streamWindowSize;
+            private long _lastWindowUpdate;
+
+            public Http2StreamWindowManager(Http2Connection connection, Http2Stream stream)
+            {
+                _connection = connection;
+                _stream = stream;
+                HttpConnectionSettings settings = connection._pool.Settings;
+                _streamWindowSize = settings._initialHttp2StreamWindowSize;
+                _windowScalingEnabled = !settings._disableDynamicHttp2WindowSizing;
+                _maxStreamWindowSize = settings._maxHttp2StreamWindowSize;
+                _windowScaleThresholdMultiplier = settings._http2StreamWindowScaleThresholdMultiplier;
+                _lastWindowUpdate = Stopwatch.GetTimestamp();
```

Comment:
It seems like this will be initialized when the Http2Stream is created. Shouldn't we wait to initialize this, until we receive response HEADERS or something like that?

Reply:
Since this is a struct, we need to initialize it anyways. Created a `Start()` method for `_lastWindowUpdate = Stopwatch.GetTimestamp();`, which is called when headers are complete.